### PR TITLE
Fix performance regression introduced by using TypeToken to resolve type variables

### PR DESCRIPTION
**Pull request by wolfs**
We now determine if there are some type variables to resolve and only resolve them in that case since using TypeToken.resolveTypeVariable seems to be very expensive.

### CODE REVIEW

File: subprojects/base-services/src/main/java/org/gradle/internal/reflect/JavaReflectionUtil.java [hunk 1]
```diff
@@ -55,4 +63,99 @@
...
+            } else if (typeToInspect instanceof GenericArrayType) {
+                GenericArrayType arrayType = (GenericArrayType) typeToInspect;
+                typesToInspect.add(arrayType.getGenericComponentType());
+            } else if (typeToInspect instanceof WildcardType) {
+                WildcardType wildcardType = (WildcardType) typeToInspect;
+                CollectionUtils.addAll(typesToInspect, wildcardType.getLowerBounds());
+                CollectionUtils.addAll(typesToInspect, wildcardType.getUpperBounds());
+            } else {
+                // We don't know what the type is - let Guava take care of it.
+                return true;
```
Comment (lptr) [inline subprojects/base-services/src/main/java/org/gradle/internal/reflect/JavaReflectionUtil.java]: So the method is only detecting if the type _might_ have a type variable?

File: subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        new TypeToken<List<BiConsumer<String, Collection<Integer>>>>() {} | false
+        new TypeToken<List<? extends Collection<String>>>() {}            | false
+        new TypeToken<List<? super List<String>>>() {}                    | false
+        new TypeToken<BiConsumer<List<String>, List<String>>>() {}        | false
+        genericReturnType("simpleGenericReturnType")                      | true
+        genericReturnType("encapsulatedTypeVariable")                     | true
+        genericReturnType("arrayTypeWithTypeVariable")                    | true
+        genericReturnType("complexTypeWithTypeVariable")                  | true
+        genericReturnType("anotherComplexTypeWithTypeVariable")           | true
+        genericReturnType("complexTypeWithArrayTypeVariable")             | true
```
Comment (lptr) [inline subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy]: Could we use Groovy's method references here (`&method`)?

File: subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        type                                                              | hasTypeVariable
+        new TypeToken<Class<?>>() {}                                      | false
+        new TypeToken<Class<String>>() {}                                 | false
+        new TypeToken<Class<String>[]>() {}                               | false
+        TypeToken.of(String.class)                                        | false
+        TypeToken.of(String[].class)                                      | false
+        new TypeToken<List<BiConsumer<String, Collection<Integer>>>>() {} | false
+        new TypeToken<List<? extends Collection<String>>>() {}            | false
+        new TypeToken<List<? super List<String>>>() {}                    | false
+        new TypeToken<BiConsumer<List<String>, List<String>>>() {}        | false
```
Comment (lptr) [inline subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy]: Why not use actual methods from `JavaReflectionUtilTestMethods` for all of these?

File: subprojects/base-services/src/main/java/org/gradle/internal/reflect/JavaReflectionUtil.java (hunk 1 above)
Comment (wolfs) [reply subprojects/base-services/src/main/java/org/gradle/internal/reflect/JavaReflectionUtil.java]: No, it detects if the type has a type variable. To my knowledge, the above cases are everything. Resolving the type variable is done by TypeToken.

I could also have thrown an exception at this point, I just though I wanted to err on the side of delegating to Guava.

File: subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy]: I don't think so. Groovy's method is a Closure, I don't know how to extract the method name again.

**Comment by wolfs** [comment]
@lptr Regarding the complexity, I guess you reference to the complexity we had before my introduction of `TypeToken`? Note that the current code supports all the possible type variables. It just does not try to resolve any type variables if there are none. `TypeToken` resolves the whole type variable table (which is probably empty) even though it does not need to resolve anything.

I think these are our options right now:
- Go back to Adam's code, and only support methods returning a type variable for code generation
- Use the optimization in this PR in some variant, i.e. only try to resolve type variables if there are any
- Implement the type variable resolution on our own in a more performant way

I would go for the optimization in this PR, though we can also go back to [Adam's original code](https://github.com/gradle/gradle/commit/b863e775563c403d0611791bd6eae3be3327826d#diff-8457bf5f65a4acd1a51614953a4261beL459). WDYT?

File: subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AbstractClassGenerator.java [hunk 4]
```diff
@@ -453,7 +454,7 @@ public ClassMetadata(Class<?> type) {
          */
         public MethodMetadata resolveTypeVariables(Method method) {
             Type returnType = method.getGenericReturnType();
-            Type resolvedReturnType = returnType instanceof Class ? returnType : TypeToken.of(type).method(method).getReturnType().getType();
+            Type resolvedReturnType = JavaReflectionUtil.hasTypeVariable(returnType) ? TypeToken.of(type).method(method).getReturnType().getType() : returnType;
```
Comment (lptr) [inline subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AbstractClassGenerator.java]: Maybe we should move all of this logic to `JavaReflectionUtil.resolveType()` or something?

**Comment by lptr** [comment]
OK, so IIUC, we only had support for `T getSomething()`, but now we have support for `Property<T> getSomething()` or `List<T> getSomething()` as well? If that's the case, all hail the PR.

**Comment by lptr** [comment]
BTW, what about something like:

```java
<X extends String> X getSomething()
```

What happens if we have something like this?

**Comment by wolfs** [comment]
> BTW, what about something like:
> 
> ```java
> <X extends String> X getSomething()
> ```
> 
> What happens if we have something like this?

Resolving doesn't do anything for the method, since the type variable is not bound to the class declaring the method. We could actually add another check for making sure the type variables which are from the declaring class of the method. If there is no type variable from the declaring class of the method, we don't need to resolve.

File: subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AbstractClassGenerator.java (hunk 4 above)
Comment (wolfs) [reply subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AbstractClassGenerator.java]: Done in https://github.com/gradle/gradle/pull/8663.

File: subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy]: Done in https://github.com/gradle/gradle/pull/8663.

File: subprojects/base-services/src/main/java/org/gradle/internal/reflect/JavaReflectionUtil.java (hunk 1 above)
Comment (lptr) [reply subprojects/base-services/src/main/java/org/gradle/internal/reflect/JavaReflectionUtil.java]: Got it. It's still weird that we just return `true` in this case, and the comment doesn't describe what's happening too well for me. I don't have a suggestion about how to make it better, though.

File: subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AbstractClassGenerator.java (hunk 4 above)
Comment (wolfs) [reply subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AbstractClassGenerator.java]: Using Guava in `JavaReflectionUtil` blows up the tooling API size: https://builds.gradle.org/viewLog.html?buildId=20597848&buildTypeId=Gradle_Check_Quick_Java11_Openjdk_Linux_toolingApi

I guess by using the `TypeToken` class we pull in quite some other things from Guava which causes the tooling API jar to increase 😞 

Let's not move the code for now, WDYT?

Comment (lptr) [reply subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AbstractClassGenerator.java]: OK

**Pull request merged**